cupy/core/fusion.pyx

```diff
@@ -895,54 +99,60 @@ class Fusion(object):
             elif isinstance(arg, complex):
                 params_info.append('D')
             else:
-                assert False
-
-        cdef tuple key = tuple(params_info)
-        if key not in self._memo:
+                raise TypeError('Unsupported input type {}.'.format(type(arg)))
+
+        cdef tuple param_key = tuple(params_info)
+        cdef tuple shape_key = tuple(shape_info)
+
+        # self._cache(dict):
+        #     key:   Tuple of dtypes and ndims of the inputs (param_key).
+        #     value: Pair of cache_shape and kernel_list.
+        # cache_shape(dict):
+        #     key:   Tuple of shapes of the inputs (shape_key).
+        #     value: Runtime kernel.
+        # kernel_list(list): List of Runtime kernel.
+        cache_shape, kernel_list = self._cache.get(param_key, (None, None))
+
+        if cache_shape is None:
+            # Initializes cache_shape and kernel_list.
+            cache_shape = dict()
+            kernel_list = []
+            self._cache[param_key] = cache_shape, kernel_list
+
+        # Find a cached kernel from cache_shape with the key as actual shape.
+        kernel, shapes = cache_shape.get(shape_key, (None, None))
+
+        if kernel is None:
+            # Find a cached kernel from kernel_list.
+            dim_map = dict()
+            for input_order, arg in enumerate(args):
+                if isinstance(arg, core.ndarray):
+                    for axis, dim in enumerate(arg.shape):
+                        dim_map[_AbstractDim(input_order, axis)] = dim
+            for cand_kernel in kernel_list:
+                if cand_kernel.shape_constraints.satisfy(dim_map):
+                    kernel = cand_kernel
+                    shapes = kernel.get_shapes_of_kernel_params(args)
+                    cache_shape[shape_key] = kernel, shapes
+                    break
+
+        if kernel is None:
+            # If not cached at all, analyze the target function.
+            history = _fusion_analysis._FusionHistory(self.name)
             try:
-                history = _FusionHistory()
                 _thread_local.history = history
-                self._memo[key] = history.get_fusion(
-                    self.func, args, self.name)
+                kernel = history.emit_kernel(self.func, args)
+                shapes = kernel.get_shapes_of_kernel_params(args)
             finally:
                 _thread_local.history = None
-        kernel, kwargs = self._memo[key]
 
-        return kernel(
-            *[a for a in args if a is not None],
-            **kwargs)
+            cache_shape[shape_key] = kernel, shapes
+            kernel_list.append(kernel)
 
-    def clear_cache(self):
-        self._memo = {}
+        return kernel.execute(args, shapes)
 
 
 def fuse(*args, **kwargs):
-    """Decorator that fuses a function.
```

Comment:
Did you remove the docstring intentionally?

Reply:
Please address this